### PR TITLE
[ops] feat: add Quack GEMM backend for fused MoE & upgrade fa4

### PR DESCRIPTION
**Pull request by Luosuu**
### What does this PR do?

1. Add [Quack kernels](https://github.com/NVIDIA/quack-kernels) (CUTLASS/CuTe DSL, SM90+) as an alternative GEMM backend for fused MoE forward/backward, alongside the existing Triton group-gemm backend. Introduces `moe_implementation: fused_quack` in `OpsImplementationConfig`, with MoE kernel patching deferred to `build_foundation_model()` instead of import time.

2. Support FA4 (`flash_attn.cute`) on Transformers v4. Previously FA4 only worked on v5 (via the hub-kernel monkey-patch). On v4, `flash_attention_forward` now passes a `SimpleNamespace` object (from `_load_veomni_local_flash_kernel`) instead of the string name, which v4's `_lazy_imports` accepts in its kernels-fallback branch via `getattr()`.

### Checklist Before Starting

- [x] Search for similar PRs: https://github.com/ByteDance-Seed/VeOmni/pulls?q=quack
- [x] PR title follows `[{modules}] {type}: {description}` format

### Test

- `pytest tests/ops/test_quack_fused_moe.py -v` — forward + backward correctness against eager reference for both split and merged fc1 weight paths, parameterized with `(8, 2, 256, 128)` and `(128, 8, 2048, 768)`.
- `pytest tests/ops/test_fused_moe_split_vs_merged.py -v` — existing Triton tests unaffected.
- Requires SM90+ GPU (H100/B200) with `quack` package installed.

### API and Usage Example

**YAML config (Quack MoE):**

```yaml
model:
  ops_implementation:
    moe_implementation: fused_quack   # or "fused" for Triton (default), "eager" for reference loop
```

**Standalone script:**

```python
model = build_foundation_model(
    config_path="...",
    moe_implementation="fused_quack",
)
```

**Direct patch (tests/benchmarks):**

```python
from veomni.ops.fused_moe import apply_veomni_fused_moe_patch
apply_veomni_fused_moe_patch(moe_implementation="fused_quack")
```

**FA4 on Transformers v4:**

```yaml
model:
  ops_implementation:
    attn_implementation: flash_attention_4  # auto-rewritten to veomni_flash_attention_4_with_sp
```

Or directly: `attn_implementation="veomni_flash_attention_4_with_sp"`.

### Design & Code Changes

**MoE Quack backend:**

- `veomni/ops/fused_moe/quack_gemm.py` — **New.** Quack GEMM backend with `QuackFusedMoeExpertFunction` (split fc1) and `MergedFc1QuackFusedMoeExpertFunction` (merged fc1). Forward/dgrad use `quack.gemm_interface.gemm` with `cu_seqlens_m`/`A_idx`; wgrad uses `cu_seqlens_k` mode.
- `veomni/arguments/arguments_types.py` — Added `fused_quack` to `moe_implementation` choices in `OpsImplementationConfig`.
- `veomni/models/auto.py` — `build_foundation_model` now accepts `moe_implementation="fused_quack"`, maps it to `config._moe_implementation="fused"` and calls `apply_veomni_fused_moe_patch(moe_implementation="fused_quack")`.
- `veomni/ops/fused_moe/__init__.py` — `apply_veomni_fused_moe_patch` now takes `moe_implementation` param to select Triton vs Quack backend. On NPU devices the parameter is ignored.
- `veomni/ops/__init__.py` — Removed MoE patching from import-time `apply_ops_patch()`.
- `veomni/utils/import_utils.py` — Added `is_quack_gemm_available()` check (SM90+).
- `tests/ops/test_quack_fused_moe.py` — **New.** Correctness tests.

**FA4 on Transformers v4:**

- `veomni/ops/flash_attn/__init__.py` — FA4 branch now passes `SimpleNamespace` object (via `_load_veomni_local_flash_kernel`) on v4 instead of the string name. On v5, the string is still used for hub-kernel interception.
- `veomni/models/auto.py` — Relaxed v5-only guard: bare `flash_attention_4` still requires v5, but `veomni_flash_attention_4_with_sp` now works on v4. Error message updated to guide users.
- `docs/design/kernel_selection.md` — Updated FA4 requirements.
- `docs/transformers_v5/veomni_flash_attention_kernel_adapter.md` — Documented FA4 v4 support mechanism and install instructions for both versions.

### Checklist Before Submitting

- [x] Applied pre-commit checks (`ruff format` + `ruff check`)
- [x] Added documentation (`docs/design/kernel_selection.md`)
- [x] Added tests (`tests/ops/test_quack_fused_moe.py`)

### E2E Loss/Grad Norm Match
image

### CODE REVIEW

File: tests/ops/test_quack_fused_moe.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        output.index_add_(0, token_idx, y.to(output.dtype))
+
+    return output
+
+
+@pytest.mark.parametrize(
+    "num_tokens,num_experts,hidden_dim,ffn_dim,topk",
+    [
+        (64, 8, 256, 128, 2),
+        (128, 128, 2048, 768, 8),
```
Comment (piyifan123) [inline tests/ops/test_quack_fused_moe.py:46]: test with more realistic model config type and longer seq?

File: veomni/utils/import_utils.py [hunk 2]
```diff
@@ -74,6 +75,15 @@ def is_fused_moe_available() -> bool:
     return torch.cuda.is_available() and not _PACKAGE_FLAGS["torch_npu"] and _PACKAGE_FLAGS["triton"]
 
 
+def is_quack_gemm_available() -> bool:
```
Comment (piyifan123) [inline veomni/utils/import_utils.py:83]: i would suggest to repurpose this function to only check availability of quack package and another function to check SM capability.

File: veomni/ops/fused_moe/__init__.py [hunk 3]
```diff
@@ -59,13 +61,28 @@ def fused_moe_forward(
...
+            On NPU devices the backend parameter is ignored and the NPU
+            kernel is always selected.
+    """
     global _fused_moe_forward
     if is_torch_npu_available():
         from .npu_group_gemm import npu_fused_moe_forward
 
         _fused_moe_forward = npu_fused_moe_forward
-    elif is_fused_moe_available() and get_env("USE_GROUP_GEMM") == "1":
+    elif backend == "quack":
```
Comment (piyifan123) [inline veomni/ops/fused_moe/__init__.py]: should we throw an error if quack package is not available or SM<90?

File: veomni/models/auto.py [hunk 4]
```diff
@@ -102,6 +103,11 @@ def build_foundation_model(
         if moe_implementation == "eager":
             logger.warning_rank0("You are using eager moe implementation, expect this to be VERY SLOW!")
 
+    from ..ops.fused_moe import apply_veomni_fused_moe_patch
```
Comment (piyifan123) [inline veomni/models/auto.py]: This sounds a little bit hacky... we should have a consolidation on how to select kernels in transformers v5 era going forward... 

we have many scattered places to monkey patch or depending on env to select kernels which is hard to maintain. 

anyway, for this PR, i think it's OK but we need a long term solution for this.

@FoolPlayer @Coach257 let's chat about this.

File: tests/ops/test_quack_fused_moe.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        fc1_1_q = fc1_1_weight.clone().detach().requires_grad_(True)
+        fc1_2_q = fc1_2_weight.clone().detach().requires_grad_(True)
+        fc2_q = fc2_weight.clone().detach().requires_grad_(True)
+
+        out_q = quack_gemm_fused_moe_forward(
+            num_experts=num_experts,
+            routing_weights=routing_weights,
+            selected_experts=selected_experts,
+            hidden_states=hs_q,
+            fc1_1_weight=fc1_1_q,
```
Comment (piyifan123) [inline tests/ops/test_quack_fused_moe.py:88]: do we need to support fc1 split mode? should we only support the new gate_up mode to simplify?

File: veomni/ops/fused_moe/quack_gemm.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+"""
+
+import torch
+from quack.gemm_interface import gemm
+
+from ...distributed.parallel_state import get_parallel_state
+from ..group_gemm.kernel.moe import expert_histogram, moe_gather, moe_scatter
+
+
+def _build_moe_indices(expert_index: torch.Tensor, num_experts: int):
```
Comment (piyifan123) [inline veomni/ops/fused_moe/quack_gemm.py:32]: add a unit test for this function since it's a little bit subtle to calculate. and a unit test with examples to assert makes it easy to understand what it does.

File: veomni/ops/fused_moe/quack_gemm.py [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+        gate_weights,
+        expert_index,
+        hidden_states,
+        fc1_1_weight,
+        fc1_2_weight,
+        fc2_weight,
+    ):
+        cu_seqlens_m, A_idx, scatter_index = _build_moe_indices(expert_index, num_experts)
+
+        # Transpose weights for forward: [E, N, K] -> [E, K, N] (view, no copy)
```
Comment (piyifan123) [inline veomni/ops/fused_moe/quack_gemm.py]: nit: add a note that quack would transpose back before calling kernels so no contiguous() call is needed.

File: veomni/ops/fused_moe/quack_gemm.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+            fc1_2_output,
+            fc1_activation,
+            scattered_gate_weight,
+            fc1_weighted_output,
+        ) = ctx.saved_tensors
+        hidden_dim = grad_output.shape[-1]
+        grad_output = grad_output.view(-1, hidden_dim)
+
+        # Step 10: scatter grad to expert-sorted order
+        grad_fc2_output = moe_scatter(grad_output, scatter_index)
```
Comment (piyifan123) [inline veomni/ops/fused_moe/quack_gemm.py:148]: so we cannot use the A_idx here for this?

File: veomni/ops/fused_moe/quack_gemm.py [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+            grad_fc2_weight = gemm(grad_fc2_output.T, fc1_weighted_output, cu_seqlens_k=cu_seqlens_m)
+
+        # Step 8-2: routing weight backward
+        grad_fc1_activation = grad_fc1_weighted_output * scattered_gate_weight
+
+        # Step 8-1: gate weight backward
+        grad_scattered_gate_weight = torch.sum(fc1_activation * grad_fc1_weighted_output, dim=-1)
+        grad_gate_weight = grad_scattered_gate_weight[scatter_index.flatten()]
+        grad_gate_weight = grad_gate_weight.reshape(gate_weights.shape)
+
```
Comment (piyifan123) [inline veomni/ops/fused_moe/quack_gemm.py:171]: should we add del in places to save vram?

File: tests/ops/test_quack_fused_moe.py (hunk 5 above)
Comment (Luosuu) [reply tests/ops/test_quack_fused_moe.py:88]: we need to support split fc1 for other libs that depends on veomni but without hf v5 upgrade

File: veomni/ops/fused_moe/quack_gemm.py (hunk 7 above)
Comment (Luosuu) [reply veomni/ops/fused_moe/quack_gemm.py]: done

File: veomni/ops/fused_moe/quack_gemm.py (hunk 9 above)
Comment (Luosuu) [reply veomni/ops/fused_moe/quack_gemm.py:171]: done

File: veomni/ops/fused_moe/quack_gemm.py (hunk 8 above)
Comment (Luosuu) [reply veomni/ops/fused_moe/quack_gemm.py:148]: grad_fc2_output is also needed in step 9 wgrad: `gemm(grad_fc2_output.T, fc1_weighted_output, cu_seqlens_k=cu_seqlens_m)`. This uses cu_seqlens_k mode where the expert-sorted dimension is K (columns of A), not M. `A_idx` only indexes into M (rows of A)

File: veomni/ops/fused_moe/quack_gemm.py (hunk 6 above)
Comment (Luosuu) [reply veomni/ops/fused_moe/quack_gemm.py:32]: sure

File: veomni/ops/fused_moe/__init__.py (hunk 3 above)
Comment (Luosuu) [reply veomni/ops/fused_moe/__init__.py]: sure

File: tests/ops/test_quack_fused_moe.py (hunk 1 above)
Comment (Luosuu) [reply tests/ops/test_quack_fused_moe.py:46]: sure

File: veomni/utils/import_utils.py (hunk 2 above)
Comment (Luosuu) [reply veomni/utils/import_utils.py:83]: sure

File: pyproject.toml [hunk 10]
```diff
@@ -244,8 +240,8 @@ flash-attn-3 = [
   { url = "https://github.com/windreamer/flash-attention3-wheels/releases/download/2026.01.12-6b9e0bf/flash_attn_3-3.0.0b1%2B20260112.cu129torch291cxx11abitrue.ea8f73-cp39-abi3-linux_x86_64.whl", marker = "extra == 'gpu'"},
 ]
 # FlashAttention 4 is developed under flash-attention/flash-attn/cute folder as a standalone python project.
-# Pinned to 02/20/2026 latest main commit.
-flash-attn-cute = { git = "https://github.com/Dao-AILab/flash-attention", subdirectory = "flash_attn/cute", rev = "6079a9bf4cfd7af8e7586afea6c49a97ebddf46e" }
+# Pinned to 03/10/2026 latest main commit.
```
Comment (piyifan123) [inline pyproject.toml:243]: can we install from pypi instead? anyway, OK to do in another PR.

Comment (Luosuu) [reply pyproject.toml:243]: I am not sure whether the pypi version support quack 0.3.2. will see.
